### PR TITLE
Clarify optional fields

### PR DESCRIPTION
**Pull request by mbohlool**
cc @thockin @janetkuo

### CODE REVIEW

File: contributors/devel/api-conventions.md [hunk 1]
```diff
@@ -579,21 +579,27 @@ Fields must be either optional or required.
...
 
-- They do not have an `omitempty` struct tag.
+- They do not have an `+optional` comment tag.
 - They are not a pointer type in the Go definition (e.g. `bool otherFlag`).
 - The API server should not allow POSTing or PUTing a resource with this field
 unset.
 
-Using the `omitempty` tag causes swagger documentation to reflect that the field
-is optional.
+Using the `+optional` (not `omitempty`) tag causes swagger documentation to reflect
```
Comment (janetkuo) [inline contributors/devel/api-conventions.md]: Should OpenAPI be mentioned somewhere in this doc?

File: contributors/devel/api-conventions.md [hunk 2]
```diff
@@ -579,21 +579,27 @@ Fields must be either optional or required.
...
 Optional fields have the following properties:
 
-- They have `omitempty` struct tag in Go.
+- They have `+optional` comment tag in Go.
 - They are a pointer type in the Go definition (e.g. `bool *awesomeFlag`) or
 have a built-in `nil` value (e.g. maps and slices).
 - The API server should allow POSTing and PUTing a resource with this field
 unset.
 
+an optional field may also include `omitempty` struct tag in go to omit empty
```
Comment (janetkuo) [inline contributors/devel/api-conventions.md]: "An"

File: contributors/devel/api-conventions.md [hunk 3]
```diff
@@ -579,21 +579,27 @@ Fields must be either optional or required.
...
+values from json serialization. If you want to have different logic for an
+optional field not provided, or provided with empty values, do not use `omitempty`.
+If you don't know what is this, and you just want an optional field, you should
+most probably has `omitempty` too.
```
Comment (janetkuo) [inline contributors/devel/api-conventions.md]: > If you want to have different logic for an optional field not provided, or provided with empty values, do not use omitempty. 

Maybe it's just me, but I find this part a bit hard to digest at a glance. Would you rephrase it?

> If you don't know what is this, and you just want an optional field, you should most probably has `omitempty` too.

Does this mean that in most cases I should add `omitempty` (in addition to `+optional`) to optional fields?

Comment (janetkuo) [inline contributors/devel/api-conventions.md]: Suggest adding a short example here for when you need or don't need to add `omitempty`.

File: contributors/devel/api-conventions.md [hunk 4]
```diff
@@ -579,21 +579,27 @@ Fields must be either optional or required.
 
 Optional fields have the following properties:
 
-- They have `omitempty` struct tag in Go.
+- They have `+optional` comment tag in Go.
```
Comment (janetkuo) [inline contributors/devel/api-conventions.md]: This applies to both internal and external types, correct?

Comment (thockin) [inline contributors/devel/api-conventions.md]: nit: "the `+optional` comment tag"

File: contributors/devel/api-conventions.md [hunk 5]
```diff
@@ -579,21 +579,27 @@ Fields must be either optional or required.
...
 - They are a pointer type in the Go definition (e.g. `bool *awesomeFlag`) or
 have a built-in `nil` value (e.g. maps and slices).
 - The API server should allow POSTing and PUTing a resource with this field
 unset.
 
+an optional field may also include `omitempty` struct tag in go to omit empty
+values from json serialization. If you want to have different logic for an
+optional field not provided, or provided with empty values, do not use `omitempty`.
```
Comment (thockin) [inline contributors/devel/api-conventions.md]: "field  which is not provided vs. provided with empty"

File: contributors/devel/api-conventions.md (hunk 2 above)
Comment (thockin) [inline contributors/devel/api-conventions.md]: "the `omitempty`"

File: contributors/devel/api-conventions.md (hunk 3 above)
Comment (thockin) [inline contributors/devel/api-conventions.md]: s/has/have

File: contributors/devel/api-conventions.md (hunk 1 above)
Comment (thockin) [inline contributors/devel/api-conventions.md]: this conflates comment tags and struct tags.  Can you clarify please?

**Comment by thockin** [comment]
Approving on principle, nits suggested

/approve

File: contributors/devel/api-conventions.md (hunk 1 above)
Comment (mbohlool) [inline contributors/devel/api-conventions.md]: Fixed.

Comment (mbohlool) [inline contributors/devel/api-conventions.md]: Fixed. Clarified that both optional and omitempty means optional for OpenAPI.

**Comment by mbohlool** [comment]
@thockin @janetkuo I've addressed your comments. PTAL.

File: contributors/devel/api-conventions.md (hunk 4 above)
Comment (janetkuo) [inline contributors/devel/api-conventions.md]: For internal types I meant something like pkg/apis/apps/types.go

**Comment by mbohlool** [comment]
Thanks. Rebased, Merged.

**Pull request merged**